
the sample_metadata.tsv file was generated by shrimp_ec23_v2.Rmd

based on TR's ResMlstCombTable50 spreadsheet

DNA Sequencing Results/MLST Results/Sample Info Added_mlst.all 
https://docs.google.com/spreadsheets/d/1H52g-yjbIhMBgJjknIDR7McCPlxu2OI72Pn0JZtH8gY/edit?usp=drive_link 
50 rows


TR 2024.0723 bmail says
The RES & MLST file should be the complete metadata file
https://docs.google.com/spreadsheets/d/1vGiYa9Dl3ydHdkMT3hOkSr34eZwSTFwlmLqJ3vmt-X0/edit?usp=drive_link
...Final Tables/Res & MLST Combined 
Took out Duplicates: total samples 44 , shrimp 10, water 34, NEW TOTAL # OF SAMPLES W/ AT LEAST ONE AMR: 36/44= 81%

I had (in shrimp_ec23.Rmd) created a join table of the 50 vs 44 
ResMlstCombTable, show 6 rows removed, but no good reason as to why they were considered dup.
The metadata between thse two files are consistent, 
so sample name  vs fasta file  vs sampling source/location  should be fine.

~~~

sample_metadata_withDupInfo.tsv is manual update of sample_metadata.tsv, added extra column for DuplicateFlag so that duplicate can be filtered.
2024.0725

The "I" sample had fastQ file, but unicycler failed to genearate fasta file, so no sequence avail.  
did not mark it as Dup 
I_CKDN230030142-1A_HGKHYDSX7


~~~

prep list for submission to NCBI SRA/GenBank

cat sample_metadata_withDupInfo.tsv | fgrep -v "Dup of" | sed 's/_Abricate_resfinder.tsv/.fasta/' > sample_metadata_noDupRows2.tsv


also remove the _L2
cat sample_metadata_withDupInfo.tsv | fgrep -v "Dup of" | sed 's/_Abricate_resfinder.tsv/.fasta/' | sed 's/_L2//' > sample_metadata_noDupRows2.tsv

to get filename as:
A_CKDN230030134-1A_HGKHYDSX7.fasta
