---
title: "analysis for Shrimp Ecuador 2023 data"
date: "2024-07-13"
output:
  html_document: default
  pdf_document: default
---



Tilden started a maper, MDPI reviwer has request...
data submission to GenBank, Niko found missing fasta for the "I" sample.

trying to reproduce some of the work.
by start "feeling" the result from mlst, etc.



write up of my findings:

* https://docs.google.com/document/d/1mP5gzfIKnpeE80osPNdE_b_Gev6pQ6_elGxjC7httpM/edit 
* https://docs.google.com/document/d/1VPa4DXD8OnTWWi78lkSQznNwNxRKxqqQGIZ_x3W-EKM/edit
* OneNotes  https://onedrive.live.com/view.aspx?resid=C00B9CE6884102BB%2190659&id=documents&wd=target%28Ecuador.one%7CE3F7F4A3-03A7-4DFD-92E8-E3BE4357E416%2FShrimp%7CA8ED9D55-1A94-4CF8-8FEA-696EA58FD115%2F%29onenote:https://d.docs.live.net/c00b9ce6884102bb/Documents/phd_prj/Ecuador.one#Shrimp&section-id={E3F7F4A3-03A7-4DFD-92E8-E3BE4357E416}&page-id={A8ED9D55-1A94-4CF8-8FEA-696EA58FD115}&end 


```{r setup}

library( pacman )
p_load( tidyverse )
p_load( dplyr )
p_load( readr )
p_load( stringr )


# fn alias 
`%nin%` = Negate(`%in%`)

# open all input files here in the setup block
# so know early on if have all necessary


####
#### read input, which is output from Abricate
####



# DATADIR="/global/scratch/users/tin/fc_graham/ecuador_2023_TJ/Sequences/Tilden/ALL/assembled-sequences_par3"  # sav
DATADIR="result/"


####
#### MLST SeqType info
####
mlst_tsv_file = sprintf( "%s/%s", DATADIR, "mlst.all.tsv") 
mlst_tsv = read_tsv( mlst_tsv_file )


####
#### resistance gene were focus of paper
####
resfind_sum_file = sprintf( "%s/%s", DATADIR, "Abricate_resfinder_summary.csv") 
resfind_sum_tmp_tsv = read_tsv( resfind_sum_file, col_types = list( .default=col_character() ) )
resfind_sum_tsv = type_convert( resfind_sum_tmp_tsv )  # here, expect only NUM_FOUND is converted as double.


resfind_sum_pct_df = resfind_sum_tsv %>%
  rename( file_name = '#FILE' ,
          RESIST_NUM_FOUND = "NUM_FOUND"
          ) 


####
##### not sure if VFDB was needed.  
####
vfdb_sum_file = sprintf( "%s/%s", DATADIR, "Abricate_vfdb_summary.csv") 
vfdb_sum_tmp_tsv = read_tsv( vfdb_sum_file, col_types = list( .default=col_character() ) )
vfdb_sum_tsv = type_convert( vfdb_sum_tmp_tsv )  # here, expect only NUM_FOUND is converted as double.


# ecoli_vf said to be vfdb + gene from literatur, so more extensive https://github.com/phac-nml/ecoli_vf
# trying it hoping as plug-in replacement to vfdb_sum_tsv
#// ecvf_sum_file = sprintf( "%s/%s", DATADIR, "ecoli_vf_summary.csv" )
#// ecvf_sum_tsv = read_tsv( ecvf_sum_file )

# Choose: VFDB vs ECVF for source of vfg
# Using VFDB

abricate_vfdb_sum_tsv = vfdb_sum_tsv    # this results 155 rows # hmmm... 124 rows in 2024.01 on wombat
#abricate_vfdb_sum_tsv = ecvf_sum_tsv   # code support this now, but result list is ~441 rows



# rest of code will be unchanged to use vfdb_sum_df
# 2024.0315, vfdb_sum_df get the jgrg result joined into it 

vfdb_sum_pct_df = abricate_vfdb_sum_tsv %>%
  rename( file_name = '#FILE' ,
          VFDB_NUM_FOUND = "NUM_FOUND"
          ) 


# View(vfdb_sum_df)

##############################################################################

# reading in files from Tilden
# shared folder:
# https://drive.google.com/drive/folders/1kk-56BYRkYOXJtascCi6LEkVxE-mO0aW
#
# files:
# https://docs.google.com/spreadsheets/d/18qKB7kIbr1adVEm0Ws56eiJPc1ZCX6T55yMw-h10PT8/edit?gid=646212993#gid=646212993
# https://docs.google.com/spreadsheets/d/15P318xrNscGkKK-ls2IUuM3HVyXZy7BhNdxrvVPiYZw/edit?gid=0#gid=0

# ref: https://docs.google.com/document/d/1mP5gzfIKnpeE80osPNdE_b_Gev6pQ6_elGxjC7httpM/edit 

CACHEDIR="CACHE"  # tmp subfolder, files not checked into git repo


SummaryTableXlsx     = sprintf( "%s/Summary Table .xlsx", CACHEDIR ) 
SummaryTableTsv      = sprintf( "%s/Summary Table  - Sheet1.tsv", CACHEDIR ) 
#SummaryTable     = readxl::read_xlsx( SummaryTableXlsx )
SummaryTable     = read_tsv( SummaryTableTsv  )  # edited: removed row 1 filename header line
# 36 rows, so not full dataset of paper

ResMlstCombTableXlsx = sprintf( "%s/Res & MLST Combined.xlsx", CACHEDIR )  
ResMlstCombTableTsv  = sprintf( "%s/Res & MLST Combined - Abricate_resfinder_summary.tsv", CACHEDIR )   # this file is in like 3 different folders, Final_Tables/, MLST Results/, Tilden_Thesis_Isolate info/ 
#ResMlstCombTable = readxl::read_xlsx( ResMlstCombTableXlsx ) %>%
ResMlstCombTable = read_tsv( ResMlstCombTableTsv ) %>%    # edited, remove tailing comments, empty rows
  rename( file_name = '#FILE' )
# 44 rows, likely full dataset of paper.  Sample "I" is here, found to have ST 3580.  0 ARG.   at least filename to sample type, farm type is correct?   ++CHEcK++

ResMlstCombTableTsv50  = sprintf( "%s/50_Sample\ Info\ Added_mlst.all\ -\ mlst.all.tsv", CACHEDIR )   
ResMlstCombTable50     = read_tsv( ResMlstCombTableTsv50 ) %>%    # edited header 
  rename( file_name = 'Filename' )
# 50 rows.  this has info before the duplicate


#### additional sample info, hoping to cross check metadata is valid
# https://docs.google.com/spreadsheets/d/1I3EjwGVHLeOC_Omm1zNeqW2yYVtEuo6X/edit?gid=774525884#gid=774525884
# [Copy of Code_DNA Extranction]  saved .tsv in metadta/ 
# albeit is is really checking if sample id has matching site info, maybe SDE (site location)

SampleInfoDnaExtractionTsv = sprintf( "%s/Copy of Code_DNA Extranction.xlsx - Sheet1.tsv", "metadata" ) 
SampleInfoDnaExtractionDf  = read_tsv( SampleInfoDnaExtractionTsv ) %>%
  rename( `SampleId` = "Column1" )

```



# create join key for  mlst  resfinder result
likely just the first 1-2 letter of the filename should suffice
eg for mlst:   (has 55 rows, sample "I" is missing)
AA_CKDN230030154-1A_HGKHYDSX7_L2.fasta
 J_CKDN230030143-1A_HGKHYDSX7_L2.fasta
2^ ^^^^3^^^^^^^^ 4^ ^5^^^^^^^ ^6 

eg for resfinder:  (has 56 rows, sample "I" is present)
I_CKDN230030142-1A_HGKHYDSX7_L2_Abricate_resfinder.tsv
2 ^^^^3^^^^^^^^ 4^ ^5^^^^^^^ ^6 7^^^^^^^^^^^^^^^^^




```{r define_join_key}


resfind_sum_pct_df = resfind_sum_pct_df %>%
  #                                                2                 3                    4                  5                      6              7=DB specific
  mutate( sample_name     = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)\\_(Abricate_resfinder).tsv"  )[,2] )%>%
  mutate( Hnum            = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)\\_(Abricate_resfinder).tsv"  )[,5] )%>%
  mutate( Lnum            = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)\\_(Abricate_resfinder).tsv"  )[,6] )%>%
  mutate( key    = str_c( sample_name, "_H", Hnum, "_L", Lnum,  sep="" ) )

# AB is not unique, ditt for AD, maybe others.


mlst_tsv_df = mlst_tsv %>%
  #                                                2                 3                    4                  5                      6              
  mutate( sample_name     = str_match( Filename, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+).fasta"  )[,2] )%>%
  mutate( Hnum            = str_match( Filename, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+).fasta"  )[,5] )%>%
  mutate( Lnum            = str_match( Filename, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+).fasta"  )[,6] )%>%
  mutate( key    = str_c( sample_name, "_H", Hnum, "_L", Lnum,  sep="" ) )
  
  

```

# join mlst with resfinder result

```{r join}

mlst_resfind_df = full_join(
  x  = mlst_tsv_df, 
  y  = resfind_sum_pct_df,
  by = "key"
  #by = c( 'Filename' = "file_name")  # nope, these filenames are not matching!  need to define a key for them.
)

#View( mlst_resfind_df )

```


# condensed tmp result, write result, this is temporarily, so can look at it using spreadsheet, for discussion

noded filenames with dup sample id
AB
AD
AH
AS
AY
BB - 6 rows... dups?
I has NA cuz the Hnum is from table w/o the ST info


```{r write_tsv}

condense_mlst_resfind_df = mlst_resfind_df %>%  
  select( sample_name.y, Hnum.x, Lnum.x, key, Filename, PubMLST_scheme_name, SeqType, RESIST_NUM_FOUND  ) 

write_tsv( condense_mlst_resfind_df, "result/condense_mlst_restfind_df.tsv")
write_tsv(          mlst_resfind_df, "result/mlst_restfind_df.tsv")

#View( condense_mlst_resfind_df )

```

########################################################

# check Tilden's result, hunting for sample "I" 

filename eg for resfinder result  ie ResMlstCombTable - Res\ \&\ MLST\ Combined\ -\ Abricate_resfinder_summary.tsv
I_CKDN230030142-1A_HGKHYDSX7_L2_Abricate_resfinder.tsv
2 ^^^^3^^^^^^^^ 4^ ^5^^^^^^^ ^6 7^^^^^^^^^^^^^^^^^


filename eg for ResMlstCombTable50 MLST Results/Sample Info Added_mlst.all 0 https://docs.google.com/spreadsheets/d/1H52g-yjbIhMBgJjknIDR7McCPlxu2OI72Pn0JZtH8gY/edit?usp=drive_link 
AA_CKDN230030154-1A_HGKHYDSX7_L2.fasta
 I_CKDN230030142-1A_HGKHYDSX7_L2_Abricate_resfinder.tsv
^2 ^^^^3^^^^^^^^ 4^ ^5^^^^^^^ ^6 7^^^^^^^^^^^^^^^^^
||                              |||||||||||||||||||
why are 27 rows  filenames end with ".fasta",  rest have the "Abricate_resfinder.tsv" ??   ++CHECK++


```{r TildenCreateJoinKey}


#SummaryTable     = readxl::read_xlsx( SummaryTableFile )
#ResMlstCombTable = readxl::read_xlsx( ResMlstCombTableFile )

#View( SummaryTable )       # use Sample ID:    ae  j 
#View( ResMlstCombTable )   # extract from Filename AE_CKDN230030158-1A_HGKHYDSX7_L2_Abricate_resfinder.tsv


#  define_join_key
ResMlstCombTable = ResMlstCombTable %>%      # version with 44 rows, ie removed DUP
  #                                                2                 3                    4                  5                      6              7=DB specific
  mutate( sample_name     = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)\\_(Abricate_resfinder).tsv"  )[,2] )%>%
  mutate( Hnum            = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)\\_(Abricate_resfinder).tsv"  )[,5] )%>%
  mutate( Lnum            = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)\\_(Abricate_resfinder).tsv"  )[,6] )%>%
  #mutate( samplename_lcase = str_to_lower( sample_name ) )  %>%
  mutate( key  = str_c( sample_name, "_H", Hnum, "_L", Lnum,  sep="" ) ) %>%
  mutate( ctrl = "ResMlstCombTable44" )

# note for version with 44 rows, ie ResMlstCombTable - Res\ \&\ MLST\ Combined\ -\ Abricate_resfinder_summary.tsv
# not unique.  ie, sample with 2 fasta file, where Hnum differs.
# AB AD AH AS  AY BB - 6 rows, check against what Tilden removed    ++CHECK++



ResMlstCombTable50 = ResMlstCombTable50 %>%
  #                                                2                 3                    4                  5                      6            
  mutate( sample_name     = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)"  )[,2] )%>%
  mutate( Hnum            = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)"  )[,5] )%>%
  mutate( Lnum            = str_match( file_name, "([:alpha:]+)[\\-_](CKDN[:digit:]+)[\\-]([:alnum:]+)[\\-_]H([:alnum:]+)DSX7[\\-_]L([:alnum:]+)"  )[,6] )%>%
  #mutate( samplename_lcase = str_to_lower( sample_name ) ) %>%
  mutate( key  = str_c( sample_name, "_H", Hnum, "_L", Lnum,  sep="" ) ) %>%
  mutate( ctrl = "ResMlstCombTable50" )

# determine which 6 seq Tilden dropped,
ResMlstCombTableBoth = full_join(
  x  = ResMlstCombTable50,
  y  = ResMlstCombTable,
  by = "key"
)  %>% select( 1:31, sample_name.y, Hnum.y, Lnum.y, ctrl.y  )
#^^ started with 44 and 50 rows, somehow end up with 51.  AB_HGLMJ_L1 is in the "removed DUP" but not in the original 50 rows!  cuz used AB_HGKHY_L2 instead.

# colnames( ResMlstCombTableBoth )

ResMlstCombTableBothCondense = ResMlstCombTableBoth  %>% 
  select(   ctrl.x, ctrl.y, site, `FARM TYPE`, source, `SAMPLE TYPE`, SDE, `SAMPLING LOCATION`, file_name.x, Species, NUM_FOUND, SeqType.x, SeqType.y, key, sample_name.x, sample_name.y, Order.x, Order.y )


SummaryTable = SummaryTable %>%
  mutate( sampleId_ucase = str_to_upper( `Sample ID:` ) )  


JoinedSumTab = full_join(
  x  = SummaryTable,
  y  = ResMlstCombTable50,
  by = c( `sampleId_ucase` = "sample_name" )
)

View( JoinedSumTab )

write_tsv(          JoinedSumTab, "CACHE/JoinedSumTab.tsv")
# may want a condensed version , strip out the genes cols... 

```

# comments on dups:
Looking at R joined tables, 44 vs 50 has these differences:
dropped 7:  AB_HGKHY_L2 AK A BC BD F L
added   1: 	AB_HGLMJ_L1

############################################
############################################



# crosscheck : Join Tilden ResMlstCombTable50   with my  parsing of the results

ResMlstCombTableBothCondense = ResMlstCombTableBoth  %>% 
  select(   ctrl.x, ctrl.y, site, `FARM TYPE`, source, `SAMPLE TYPE`, SDE, `SAMPLING LOCATION`, file_name.x, Species, NUM_FOUND, SeqType.x, SeqType.y, key, sample_name.x, sample_name.y, Order.x, Order.y )


```{r tinVsTilden}


TinVsTilden = full_join(
  x  = ResMlstCombTableBoth,               # choices: ResMlstCombTable50 ResMlstCombTableBoth ResMlstCombTableBothCondense
  y  = mlst_resfind_df,           # choices: condense_mlst_resfind_df, mlst_resfind_df
  by = "key"
  #by = c( `sample_name.y` = "sample_name" )
)

# TinVsTilden has 56 rows.   she gave me 56 fasta files, Abricate result indicate 56 outputs. 
# sample I_HGKHY_L2 is mostly NA.
# 6 rows not in Tilden's ResMlstCombTable50, which are rows where there were dup FASTA file for samples (multiple H/L num fasta files): 
# AB AD AH AS  AY BB 

# so Tilden did trim those, got to 50
# and removed 6 rows again cuz she found other dups?

colnames( TinVsTilden )

# 56 rows in mine
# 44 rows in Tilden (the rm Dup version (.y?))
TinVsTildenCompareST = TinVsTilden %>%
  select( Filename, key, sample_name.x.x, sample_name.y.y, SeqType, SeqType.x, SeqType.y,  ctrl.x, ctrl.y, site, `FARM TYPE`, source, `SAMPLE TYPE`, SDE, `SAMPLING LOCATION`, file_name.x, Species,  Order.x, Order.y, RESIST_NUM_FOUND, NUM_FOUND  )


TinVsTildenCompareAMR = TinVsTilden %>%
  select( key, sample_name.x.x, sample_name.y.y, RESIST_NUM_FOUND, NUM_FOUND, Species,  Order.x, Order.y ,  ctrl.x, ctrl.y, site, `FARM TYPE`, source, `SAMPLE TYPE`, SDE, `SAMPLING LOCATION`, Filename, file_name.x )


#### next see if can get more than 44 rows, at least find a replacement for sample "I"
#### maybe able to get 50, sample Z need to check.  AB has two versions of fasta... 

TinVsTildenCondense = TinVsTilden %>%
  select( 1:30, sample_name.x.x, sample_name.y.y, ctrl.y, - Hnum.x.x, - Lnum.x.x   )
#^^ started with 44 and 50 rows, somehow end up with 51.  AB_HGLMJ_L1 is in the "removed DUP" but not in the original 50 rows!  cuz used AB_HGKHY_L2 instead.


```

# observations TinVsTildenCompareST
SeqType mimatch (at lease BB was mismatched in Tilden's 50 vs 44):
B BA  BB 
I  (i don't have, she think it is 3580)
J  ( I have ST3580 for this)
K ( 6186, what Niko thought was a shifted row?)
Q
R S T W 
AD
AS AY BB
Z, AB, A, F, L are martial mismatch


when in doubt, use mine, cuz mine are joined programaticaly, she might have done it by hand?

# observations TinVsTildenCompareAMR
RESIST_NUM_FOUND match for subset of 44 rows in common.
Sample "I" RESIST_NUM_FOUND=0  (cuz Abricate unable to process) .  
Sample "I" likely should be dropped cuz no ST.   

Use subset 44 ?? or 50?  or subset that has data?

Probably around 49, cuz take out sample I for lack of fasta,
but don't otherwise have reasons to believe there are dups ...
cuz her original list has 56, 6 were removed for reasons noted
and I removed dups from the H/L num that otherwise is for the same sample id.




############################################

# cross check sample metadata
limited abilities to check, but hopeping to get at least some reassurance SampleID and location match

```{r cross_chk_sample_metadata }
SampleMetadataXchk2 = full_join(
  x  = SampleInfoDnaExtractionDf,
  y  = TinVsTildenCompareAMR,           # choices: TinVsTilden
  by = c( `SampleId` = "sample_name.x.x" )
)

SampleMetadataXchk2condensed = SampleMetadataXchk2 %>%
  select( `Notes:`, SampleId, sample_name.y.y, Column2, site, `FARM TYPE`, SDE, `SAMPLING LOCATION`, source, `SAMPLE TYPE`, )

# View( SampleMetadataXchk2 )
View( SampleMetadataXchk2condensed )


# observations
# mismatch
# AC "3C, D (3), 5/30"    vs 1C   WATER
# AB "1C, E (2), 5/30"    vs 3C   WATER "multiple becteria taken from same plate"
# maybe just flipped, both are WATER, unlikely to change result.  ie a conservative swap :)

```


############################################
############################################


# new result, ie version 2
start from my table, 
only pull meta data from Tilden
expect about 50 rows
see if there are dups that may want to trim down to ~44.

```{r resultVer2 }


#newResut = TinVsTilden %>%
#  select( Filename, key, sample_name.x.x, SeqType, SeqType.x, SeqType.y,  ctrl.x, ctrl.y, RESIST_NUM_FOUND, NUM_FOUND, Species, source, `SDE`   )

metadata = ResMlstCombTable50 %>%
  select( key, ctrl,  date, site, source, SDE, file_name, sample_name, Hnum, Lnum )

resultVer2 = full_join(
  x  = metadata, 
  y  = mlst_resfind_df,    # Tin's result
  by = "key"
) 
# %>%   select( -Hnum, -Lnum )
  
  
  

#  select( key, ctrl,  Allele_ID_1, Allele_ID_2, Allele_ID_3, Allele_ID_4, Allele_ID_5, Allele_ID_6, Allele_ID_7,  sample_name, Hnum, Lnum,  )
  
  
  
# after done, then maybe join with previuos compare table and see differences...
  
```


############################################
############################################

# create new plot 
Reviewer #4 disliked the pie chart of Figure 4.  recommended stacked bar graph instead.
However, Tilden has 34 samples from water, 10 from shrimp.
Not sure which samples were selected.
The joined table above have ~55 rows.



# Notes about sequence files

most fasta files are about 480 - 550 MB.  But a few are 1210 MB!  ls -lS (it sorted by size, largest first)

-rw-rw-r-- 1 tin users 12115595 Nov 15  2023 AK_CKDN230030164-1A_HGKHYDSX7_L2.fasta
-rw-rw-r-- 1 tin users 10740528 Nov 15  2023 AU_CKDN230030174-1A_HGKHYDSX7_L2.fasta
-rw-rw-r-- 1 tin users  9661854 Nov 15  2023 BA_CKDN230030180-1A_HGKHYDSX7_L2.fasta
-rw-rw-r-- 1 tin users  9033176 Nov 15  2023 S_CKDN230030149-1A_HGKHYDSX7_L2.fasta
-rw-rw-r-- 1 tin users  7035141 Nov 15  2023 E_CKDN230030138-1A_HGKHYDSX7_L2.fasta
-rw-rw-r-- 1 tin users  6635702 Nov 15  2023 AE_CKDN230030158-1A_HGKHYDSX7_L2.fasta
-rw-rw-r-- 1 tin users  6448173 Nov 15  2023 AV_CKDN230030175-1A_HGKHYDSX7_L2.fasta
-rw-rw-r-- 1 tin users  5913388 Nov 15  2023 AZ_CKDN230030179-1A_HGKHYDSX7_L2.fasta
-rw-rw-r-- 1 tin users  5693370 Nov 15  2023 F_CKDN230030139-1A_HGKHYDSX7_L2.fasta
...
-rw-rw-r-- 1 tin users  4505363 Nov 15  2023 Z_CKDN230030153-1A_HGKHYDSX7_L2.fasta





